mathesar_ui/src/stores/table-data/constraints.ts

```diff
@@ -95,6 +120,92 @@ export class ConstraintsDataStore implements Writable<ConstraintsData> {
     return null;
   }
 
+  async add(constraintDetails: Partial<Constraint>): Promise<Partial<Constraint>> {
+    const constraint = await this.api.add(constraintDetails);
+    await this.fetch();
+    return constraint;
+  }
+
+  async remove(constraintId: number): Promise<void> {
+    await this.api.remove(constraintId);
+    await this.fetch();
+  }
+
+  /**
+   * A constraint only matches if the set of its columns strictly equals the set
+   * of columns supplied here. For example, if a constraint is set on three
+   * columns, two of which are passed to this function, that constraint will
+   * _not_ be returned.
+   */
+  constraintsThatMatchSetOfColumns(columns: Column[]): Readable<Constraint[]> {
+    const columnsNames = columns.map((c) => c.name);
+    function isMatch(constraint: Constraint) {
+      if (constraint.columns.length !== columnsNames.length) {
+        return false;
+      }
+      return constraint.columns.every(
+        (constraintColumn) => columnsNames.includes(constraintColumn),
+      );
+    }
+    return derived(this.store, (s) => s.constraints.filter(isMatch));
+  }
+
+  /**
+   * Caveat: even though primary key columns must be unique, this function will
+   * give `false` for them because they don't usually have unique constraints
+   * set too.
+   */
+  columnHasUniqueConstraint(column: Column): Readable<boolean> {
+    const constraints = this.constraintsThatMatchSetOfColumns([column]);
+    return derived(constraints, (c) => c.some((constraint) => constraint.type === 'unique'));
+  }
+
+  columnAllowsDuplicates(column: Column): Readable<boolean> {
+    return derived(
+      this.columnHasUniqueConstraint(column),
+      (hasUniqueConstraint) => {
+        if (column.primary_key) {
+          return false;
+        }
+        return !hasUniqueConstraint;
+      },
+    );
+  }
```

Comment:
As mentioned in the previous comments, creating a new store and returning it here is not optimal.

My suggestion would be to:
* Create just one derived store as a property of ConstraintsDataStore, called duplicatesAllowingColumns or similar. We can think of a better name.

  ```
   type DuplicatesAllowingColumnsSubstance = Set<Column['name']>
   type DuplicatesAllowingColumnsStore = Readable<DuplicatesAllowingColumnsSubstance>
   ```
* This store would only contain columns that allow duplicates. This will include non-unique, non-primarykey columns.
* Within the component, we'd just do `$duplicatesAllowingColumns.has(column.name)`